eeps/eep-0054.md

```diff
@@ -0,0 +1,346 @@
+arguments to the call were incorrect in some way (in this example,
+both arguments have wrong types).
+
+We propose a mechanism that enables the shell to print more helpful
+error messages.  Here is how the message would be printed with the
+reference implementation of this EEP:
+
+    1> element(a, b).
+    ** exception error: bad argument
+         in function  element/2
+            called as element(a,b)
+            *** argument 1: not an integer
+            *** argument 2: not a tuple
+
+Note that the exact formatting and phrasing of the messages is an implementation
+detail outside the scope of this EEP.  What will be specified here are the APIs
+and conventions that make these messages possible.
+
+
+### Proposals in this EEP
+
+* An extension of the format of the call-stack back trace (**stacktrace**) format
+  to indicate that there exists extended error information for that call, and
+  a convention for a extended error information can be provided.
+
+* A new `erlang:error/3` BIF to allow libraries and applications to
+  raise an exception with extended error information.
+
+* New functions `erl_error:format_exception/3` and
+  `erl_error:format_exception/4` to allow libraries and application to
+  format stacktraces in the same style as the shell.
+
+
+### Extending the stacktrace
+
+The stack back-trace (**stacktrace**) is currently a list of tuples.
+For the purpose of this EEP we are only interested in the first entry
+in the stacktrace.  It looks like
+`{Module,Function,Arguments,ExtraInfo}`, where `ExtraInfo` is a list of
+two-tuples.
+
+As an indication that extended error info is available, we propose
+adding an `{error_info,ErrorInfo}` tuple to `ExtraInfo` in the
+first element in the stacktrace.  If that tuple is present when an
+exception is caught, it indicates that `Module` has a
+`format_error/3` function that can be called to provide additional
+information about the error.  The arguments for `format_error/3` are
+`Function`, `Arguments`, and `ErrorInfo` from the `error_info` tuple.
+
+Thus, if the first entry in the stacktrace is:
+
+    {erlang,element,[1,no_tuple],[{error_info,ErrorInfo}]}
+
+the following call should be made to provide additional information
+about the error:
+
+    erlang:format_error(element, [1,no_tuple], ErrorInfo)
+
+The `format_error/3` function should return a list of **unicode:chardata()**
+values.  The first element in the list describes the error in the
+first argument, the second element the error in the second argument,
+and so on.  If an element is the empty list, it means that there is no
+error information for that argument.
+
+As an example:
+
+    erlang:format_error(element, [1,no_tuple], ErrorInfo)
+
+could return:
+
+    [[], <<"not a tuple">>]
+
+And:
+
+    erlang:format_error(element, [99,{a,b,c}], ErrorInfo)
+
+could return:
+
+    [<<"out of range">>]
+
+or:
+
+    [<<"out of range">>, []]
+
+Note that the `ErrorInfo` term is only to be used by `Module:format/3`.
```

Comment:
should that `Module:format/3` ref be to `Module:format_error/3` ?

Reply:
Yes, it should.